Refactored Skopeo-Copy and Authentication done via Secret and Service Account

anandxkumar opened the pull request:
This PR is in reference to the comment (https://github.com/openshift-pipelines/task-containers/pull/3#discussion_r1148954825)

In this PR, I have introduced authentication via `secret` and `service account`. I have added a new `secret.yaml`  The secret.yaml creates a `sc-secret` secret object and `sc-account` service account that refers to `sc-secret`.

In the task defintion, I have added volume  `registry-secret` that refers to `sc-secret` and mount this volume in the copy task. I'm using `$(cat /var/run/secrets/sc-secret/src_username)` to basically get the secret value, I tried other methods but was unsuccessfuly, let me know if there is a better way to do it.

I refactored the basic layout of the Skopeo-copy with params, inorder to make it more modular, then just having a single parameter for getting the URL:

- `SOURCE_REGISTRY`
- `SOURCE_IMAGE`
- `SOURCE_TAG`
- `DESTINATION_REGISTRY`
- `DESTINATION_IMAGE`
- `DESTINATION_TAG`

Let me know if you think it is a better approach or I should switch back to the previous version. 
I think after we finalize the basic layout of the task and authentication, we can proceed ahead with the testing part (https://github.com/openshift-pipelines/task-containers/pull/3)

otaviof commented on templates/secret.yaml:
```diff
@@ -0,0 +1,19 @@
+apiVersion: v1
+kind: Secret
+metadata:
+  name: sc-secret
+  annotations:
+    tekton.dev/docker-0: https://docker.io
+type: kubernetes.io/basic-auth
+stringData:
+  src_username: <>
+  src_password: <>
+  dest_username: <>
+  dest_password: <>
```
Instead of having this Secret part of the templates, we should have a `kubectl` example. For instance [here](https://github.com/otaviof/task-git/blob/main/README.md#basic-auth-httpssh), however the Secret type is `--type=kubernetes.io/dockerconfigjson`.

otaviof commented on templates/secret.yaml:
```diff
@@ -0,0 +1,19 @@
+  src_password: <>
+  dest_username: <>
+  dest_password: <>
+---
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: sc-account
+secrets:
+  - name: sc-secret
```
Here we should mention the service account needs to be edited (`kubectl edit serviceaccount ...`), you can show a example on `README.md`.

otaviof commented on templates/taskrun-sc.yaml:
```diff
@@ -0,0 +1,22 @@
+apiVersion: tekton.dev/v1beta1
+kind: TaskRun
+metadata:
+  name: skopeo-copy-run
+spec:
+  serviceAccountName: sc-account
+  taskRef:
+    name: skopeo-copy
+  params:
+    - name: SOURCE_REGISTRY
+      value: <>             # eg - docker.io
+    - name: SOURCE_IMAGE
+      value: <>             # username/imageName
+    - name: SOURCE_TAG
+      value: latest         # By default
+    - name: DESTINATION_REGISTRY
+      value: <>
+    - name: DESTINATION_IMAGE
+      value: <>
+    - name: DESTINATION_TAG
+      value: latest
+
```
The `TaskRun` is an ephemeral object, should be only issued when the Task is used. Please remove the whole file from the templates.

otaviof reviewed (changes requested):
Please consider the review comments, what we are looking forward here is having the traditional Tekton Pipeline authentication, which mostly consists on the documentation example.

anandxkumar commented:
@otaviof, I tried to create Service-account using 
```
kubectl create serviceaccount my-service-account
kubectl patch serviceaccount my-service-account -p '{"imagePullSecrets": [{"name": "secret-key"}]}'
```
But that failed during authentication, so I have added the service-account template, and added instructions in the Readme about how to edit it. Do you think it's a good approach to have the service account template?

otaviof commented:
> @otaviof, I tried to create Service-account using
> 
> ```
> kubectl create serviceaccount my-service-account
> kubectl patch serviceaccount my-service-account -p '{"imagePullSecrets": [{"name": "secret-key"}]}'
> ```
> 
> But that failed during authentication, so I have added the service-account template, and added instructions in the Readme about how to edit it. Do you think it's a good approach to have the service account template?

The service account will be provided to the Task by a `TaskRun`/`PipelineRun` instance, we only need to document some examples how to configure the authentication. The `kubectl patch` above is a good example showing how to configure the service account, which comes right after a `kubectl create secret --type=...` example.

Bare in mind both secret and service account will most likely be already created, we only need to document the configuration requirements, linking users with upstream official documentation.

To troubleshoot the authentication, make sure the service account being used is the same where you included the pull-secret.

otaviof commented on README.md:
```diff
@@ -6,12 +6,12 @@ Skopeo is a command line tool for working with remote image registries.
       kubectl create secret generic docker-registry \
-      --from-file=.dockerconfigjson=<path/to/.docker/config.json> \
-      --type=kubernetes.io/dockerconfigjson
+            --from-file=.dockerconfigjson=${HOME}/.docker/config.json \
```
Let's leave only method "2" as the preferred example, please.

anandxkumar replied:
So I should completely remove Method 1?

otaviof commented on README.md:
````diff
@@ -29,23 +29,37 @@ The copy command will take care of copying the image from internal.registry to p
         }
       }
       ```
+      To setup via method 1, you can refer to [the official kubernetes documentation](https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/#log-in-to-docker-hub). 
````
Here we should link Tekton's documentation: https://tekton.dev/docs/pipelines/auth/

otaviof commented on README.md:
```diff
@@ -29,23 +29,37 @@ The copy command will take care of copying the image from internal.registry to p
-To create the service account one can use the template provided in `templates/service-account.yaml`, edit the secrets and service account name accordingly and run <br />`$ kubectl apply -f templates/service-account.yaml` 
+      To setup via method 2, you can refer to [the official kubernetes documentation](https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/#log-in-to-docker-hub).  
+
+      <br />
```
Please, avoid using HTML tags on Markdown, remove the `<br />` 🙇‍♀️

otaviof reviewed (changes requested):
Cool, good progress here!

otaviof reviewed (approved):
/lgtm

Cool, good progress here! Thanks.
